flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/DataTypeFactory.java

```diff
@@ -0,0 +1,110 @@
+ */
+
+package org.apache.flink.table.catalog;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.table.annotation.DataTypeHint;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.api.ValidationException;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.extraction.DataTypeExtractor;
+import org.apache.flink.table.types.logical.DistinctType;
+import org.apache.flink.table.types.logical.StructuredType;
+
+import java.util.Optional;
+
+/**
+ * Factory for creating fully resolved data types that can be used for planning.
+ *
+ * <p>The factory is useful for types that cannot be created with one of the static methods in
+ * {@link DataTypes}) because they require access to configuration or catalog.
+ */
+@PublicEvolving
+public interface DataTypeFactory {
+
+	/**
+	 * Creates a type by a fully or partially defined name.
+	 *
+	 * <p>The factory will parse and resolve the name of a type to a {@link DataType}. This includes
+	 * both built-in types as well as user-defined types (see {@link DistinctType} and {@link StructuredType}).
+	 */
+	Optional<DataType> createDataType(String name);
+
+	/**
+	 * Creates a type by a fully or partially defined identifier.
+	 *
+	 * <p>The factory will parse and resolve the name of a type to a {@link DataType}. This includes
+	 * both built-in types as well as user-defined types (see {@link DistinctType} and {@link StructuredType}).
+	 */
+	Optional<DataType> createDataType(UnresolvedIdentifier identifier);
+
+	/**
+	 * Creates a type by analyzing the given class.
+	 *
+	 * <p>It does this by using Java reflection which can be supported by {@link DataTypeHint} annotations
+	 * for nested, structured types.
+	 *
+	 * <p>It will throw an {@link ValidationException} in cases where the reflective extraction needs
+	 * more information or simply fails.
+	 *
+	 * <p>The following examples show how to use and enrich the extraction process:
+	 *
+	 * <pre>
+	 * {@code
+	 *   // returns INT
+	 *   createDataType(Integer.class)
+	 *
+	 *   // returns TIMESTAMP(9)
+	 *   createDataType(java.time.LocalDateTime.class)
+	 *
+	 *   // returns an anonymous, unregistered structured type
+	 *   // that is deeply integrated into the API compared to opaque RAW types
+	 *   class User {
+	 *
+	 *     // extract fields automatically
+	 *     public String name;
+	 *     public int age;
+	 *
+	 *     // enrich the extraction with precision information
+	 *     public @DataTypeHint("DECIMAL(10,2)") BigDecimal accountBalance;
+	 *
+	 *     // enrich the extraction with forcing using RAW types
+	 *     public @DataTypeHint(forceRawPattern = "scala.") Address address;
+	 *
+	 *     // enrich the extraction by specifying defaults
+	 *     public @DataTypeHint(defaultSecondPrecision = 3) Log log;
+	 *   }
+	 * }
+	 * </pre>
+	 */
+	default <T> DataType createDataType(Class<T> clazz) {
```

Comment:
Do we need that default method? I don't think we will ever have a multiple implementations of that interface anyway, so one could implement everything here.